api/tracer_test.go

```diff
@@ -0,0 +1,118 @@
+package api
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/gobuffalo/uuid"
+	"github.com/netlify/gotrue/conf"
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type TracerTestSuite struct {
+	suite.Suite
+	API    *API
+	Config *conf.Configuration
+
+	instanceID uuid.UUID
+}
+
+func TestTracer(t *testing.T) {
+	api, config, instanceID, err := setupAPIForTestForInstance()
+	require.NoError(t, err)
+
+	ts := &TracerTestSuite{
+		API:        api,
+		Config:     config,
+		instanceID: instanceID,
+	}
+	defer api.db.Close()
+
+	suite.Run(t, ts)
+}
+
+func (ts *TracerTestSuite) TestTracer_Spans() {
+	spans := []string{}
+
+	tt := &testTracer{spans: &spans, tags: &map[string]string{}}
+	opentracing.SetGlobalTracer(tt)
+
+	w := httptest.NewRecorder()
+	req := httptest.NewRequest(http.MethodPost, "http://localhost/something1", nil)
+	ts.API.handler.ServeHTTP(w, req)
+	req = httptest.NewRequest(http.MethodPost, "http://localhost/something2", nil)
+	ts.API.handler.ServeHTTP(w, req)
+
+	assert.Equal(ts.T(), http.StatusNotFound, w.Code)
+	assert.Equal(ts.T(), spans, []string{"http.handler", "http.handler"})
+}
+
+func (ts *TracerTestSuite) TestTracer_Tags() {
+	tags := map[string]string{}
+
+	tt := &testTracer{tags: &tags, spans: &[]string{}}
+	opentracing.SetGlobalTracer(tt)
+
+	w := httptest.NewRecorder()
+	req := httptest.NewRequest(http.MethodPost, "http://localhost/something", nil)
+	ts.API.handler.ServeHTTP(w, req)
+
+	assert.Equal(ts.T(), http.StatusNotFound, w.Code)
+	assert.Equal(ts.T(), tags["http.method"], "POST")
+	assert.Equal(ts.T(), tags["http.status_code"], "404")
+	assert.Equal(ts.T(), tags["http.url"], "/something")
+	assert.Equal(ts.T(), tags["resource.name"], "POST /something")
+	assert.NotEmpty(ts.T(), tags["http.request_id"])
+}
+
+type testTracer struct {
+	noopTracer opentracing.NoopTracer
```

Comment:
if you make this an embedded struct field you don't have to implement `Inject` and `Extract`